.github/workflows/test-environment.yml

```diff
@@ -0,0 +1,223 @@
+          echo 'KUBE_CONFIG_DATA=$(cat ~/.kube/config | base64)' >> $GITHUB_ENV
+          aws s3 cp ~/.kube/config "${{ env.S3_BUCKET }}/kubeconfig"
+          kubectl config use-context eks-config
+          kubectl apply -f ${{ env.FLEET_API_DIR}}/kspm_eks.yaml
+
+      - name: Install KSPM Unmanaged integration
+        id: kspm-unmanaged
+        run: |
+          cd fleet_api
+          poetry run python src/install_kspm_unmanaged_integration.py
+
+      - name: Deploy KSPM Unmanaged agent
+        run: |
+          chmod 600 ${{ env.EC2_KSPM_KEY }}
+          # Copy the manifest file to the EC2 instance
+          scp -o StrictHostKeyChecking=no -v -i ${{ env.EC2_KSPM_KEY }} ${{ env.FLEET_API_DIR}}/kspm_unmanaged.yaml "ubuntu@${{ env.KSPM_PUBLIC_IP }}:~/."
+          # Apply the manifest file
+          ssh -o StrictHostKeyChecking=no -v -i ${{ env.EC2_KSPM_KEY }} "ubuntu@${{ env.KSPM_PUBLIC_IP }}" "kubectl apply -f kspm_unmanaged.yaml"
+
+      - name: Install CSPM integration
+        id: cspm
+        run: |
+          cd fleet_api
+          poetry run python src/install_cspm_integration.py
+
+      - name: Deploy CSPM agent
+        run: |
+          chmod 600 ${{ env.EC2_CSPM_KEY }}
+          # Copy the manifest file to the EC2 instance
+          scp -o StrictHostKeyChecking=no -v -i ${{ env.EC2_CSPM_KEY }} ${{ env.FLEET_API_DIR}}/cspm-linux.sh "ubuntu@${{ env.CSPM_PUBLIC_IP }}:~/."
+          # Apply the manifest file
+          ssh -o StrictHostKeyChecking=no -v -i ${{ env.EC2_CSPM_KEY }} "ubuntu@${{ env.CSPM_PUBLIC_IP }}" "chmod +x cspm-linux.sh && ./cspm-linux.sh"
+
+      - name: Upload Integrations data
+        if: always()
+        env:
+          S3_BUCKET: "${{ env.S3_BASE_BUCKET }}/${{ env.DEPLOYMENT_NAME }}_${{ env.TF_STATE_FOLDER }}"
+        run: |
+          aws s3 cp "${{ env.FLEET_API_DIR}}/kspm_unmanaged.yaml" "${{ env.S3_BUCKET }}/kspm_unmanaged.yaml"
+          aws s3 cp "${{ env.FLEET_API_DIR}}/kspm_eks.yaml" "${{ env.S3_BUCKET }}/kspm_eks.yaml"
+          aws s3 cp "${{ env.FLEET_API_DIR}}/cspm-linux.sh" "${{ env.S3_BUCKET }}/cspm-linux.sh"
+
+      # Placeholder for sanity tests
+      # - name: Run Sanity tests
+      #   if: success()
+      #   run: |
+      #     sleep 300
+
+      - name: Summary
+        if: always()
+        run: |
+          kibana_url=$(terraform output -raw kibana_url)
+          summary="Kibana URL: $kibana_url"
+          bucket_name="${{ env.S3_BASE_BUCKET }}"
+          bucket_name="${bucket_name#s3://}"
+          s3_bucket_link="[creds and keys](https://s3.console.aws.amazon.com/s3/buckets/$bucket_name)"
+          summary=$(cat <<-EOF
+          Kibana URL: [kibana]($kibana_url)
+          Environment Details: $s3_bucket_link
+          EOF
+          )
+          echo "$summary" >> $GITHUB_STEP_SUMMARY
+          echo "$summary" # Print the summary to the workflow log
+
+      - name: Cleanup Environment
+        if: github.event.inputs.cleanup-env == 'true'
```

Comment:
Do you think it makes sense to also do the cleanup if some of the stages failed?

Reply:
Developing the logic for that might be a bit complicated. It depends on the nature of the failure—if the failed steps are unrelated to resource provisioning, it might not be necessary to perform cleanup. In such cases, the environment itself might be in good condition, and the issue could be related to the installation of integrations or the execution of tests. Personally, I believe it's better to start with a simpler approach and improve the action as we encounter specific problems along the way.

Reply:
would doing a cleanup anyway could lead to any issues? if not, why not run it on every failure?

Reply:
Running a cleanup process on every failure may not be necessary or relevant in certain cases. For example, if the environment provisioning was successful but there were issues with installing agent policies or other components, performing a cleanup might not be the appropriate action.